### ISSUE TITLE
suppression directives in unreachable code are not ignored

### ISSUE THREAD

**Issue by pantheraleo-7** [astral-sh/ty#3706]
### Summary

```python
import sys


def f(_: int): ...


if sys.platform != sys.platform:
    f("")  # ty: ignore
```

gives the following warning:

```
Unused blanket `ty: ignore` directive

help: Remove the unused suppression comment
```

the directive _is_ used when the code is reachable, so the `unused-ignore-comment` warning is a false positive.

---

strangely though, this does not happen in all cases:

```python
import sys

if sys.platform != sys.platform:
    a: int = ""  # ty: ignore
```

this works just fine.

### Version

v0.0.44

**Comment by MichaReiser** [comment]
This is related to https://github.com/astral-sh/ty/issues/2681. But I'd suggest we keep this open because it specifically points out the inconsistency around assignments.

**Comment by carljm** [comment]
ty does not blanket-suppress all diagnostics in unreachable code, only ones which depend on definitions from outside the unreachable region. So in the first example, we treat `f` as having type `Never`, since no possible definition of `f` could reach this call. (This seems trivial in this example where there is only one definition of `f`, but in more complex examples where e.g. we check a version conditional and the definition of `f` is also version-conditioned, this is pretty important.)

In the second example, `a: int = ""` is a type error regardless of any external type information, so we still emit that diagnostic.

I'm not sure that keeping this issue open really helps? I think the resolution to #2681 should be to suppress `unused-ignore-comment` in unreachable code. The fact that ty can still sometimes emit diagnostics in unreachable code is not really relevant; those suppression comments should continue to work.

**Comment by pantheraleo-7** [comment]
> ty does not blanket-suppress all diagnostics in unreachable code

which is correct. this issue was asking to suppress the `unused-ignore-comment` warning in unreachable code. so this issue is a 1:1 duplicate of https://github.com/astral-sh/ty/issues/2681, which didn't came up in my search coz I didn't search by the rule name. sorry for that (I was in the middle of something).

**Comment by carljm** [comment]
Yes, understood. My comment was explaining the "strangely, though..." part of your report. (Why the one type error still requires a suppression comment in u reachable code while the other does not.)

**Comment by pantheraleo-7** [comment]
> requires a suppression comment in u(n)reachable code

uhh.. no? both the snippets do _not_ require a suppression comment in the unreachable case. the comment is put there for the reachable case. the problem is that ty flags the comment as unused in the unreachable case for snippet 1 but not snippet 2.

---

without ignore comments, no diagnostics. even for the type errors [coz they're unreachable]:
https://play.ty.dev/fe9c15db-a071-4790-bf0c-147ccd357868

with ignore comments, the `unused-ignore-comment` warning is the sole diagnostic. but only for case 1:
https://play.ty.dev/9c3f2e73-c2c4-4ade-a38a-c8649d01e0ab

**Comment by carljm** [comment]
Oh hm, interesting, thanks. It looks like we [changed our behavior](https://github.com/astral-sh/ruff/pull/24179) to suppress all diagnostics in unreachable code. And the mechanism we used there results in this inconsistency; the `f("")` diagnostic never occurs at all (since `f` is typed as `Never`), whereas the `a: int = ""` diagnostic does occur, and then is silenced by the suppression comment before it hits the reachability filter, avoiding `unused-ignore-comment`.

Ultimately I think this inconsistency still won't matter, if we also suppress all `unused-ignore-comment` in unreachable code.

**Comment by pantheraleo-7** [comment]
my 2 cents:

I think we should only blanket-suppress code that is unreachable for all cases - which is a linting error anyway.

for code that is unreachable for the current platform or settings, but is valid code somewhere somehow. it should be muted. but also checked and reported in a way as if it were run on the reachable case.

for ex, if I have `if sys.platform=="win32": ...` on macOS, it would be really nice to get diagnostics as Windows platform. I think doing this for OS platforms and python versions makes sense(?) and would be a good MVP.

For cross-platform codebases, this would be a killer feature. ty is already better than others[^1] as it provides some level of on-hover support inside unreachable code blocks (hoping for full support once https://github.com/astral-sh/ty/issues/2272 resolves).

[^1]: zuban reports each and everything as `Any`. pyrefly just says "unreachable code".

**Comment by carljm** [comment]
Distinguishing "never reachable" from "not reachable due to version/platform settings" is quite tricky in general, though plausible that a simplified version could be "good enough". See discussion in https://github.com/astral-sh/ty/issues/1948 and https://github.com/astral-sh/ty/issues/3312.

We originally wanted to do true multi-version and multi-platform checking in a single pass (see #160 and #1309) but it is complex enough that we had to defer it in favor of more pressing features. Would love to come back to it sometime.
